### PR TITLE
Update WebAssembly Document (WASM.md).

### PR DESCRIPTION
**Pull request by jplevyak**
Signed-off-by: John Plevyak <jplevyak@gmail.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: WASM.md [hunk 1]
```diff
@@ -1,23 +1,64 @@
...
+. ./emsdk\_env.sh
+
+It is possible later versions will work, e.g.
+./emsdk install latest
+./emsdk activate latest
+However 1.38.25 is known to work.
+
+## clang-7 or clang-8
+
+export CC=clang
```
Comment (PiotrSikora) [inline WASM.md]: For this to work, you usually need to export `PATH` as well.

File: WASM.md [hunk 2]
```diff
@@ -1,23 +1,64 @@
...
+export CXX=clang++
+
+## tools
+
+sudo apt-get install autoconf automake libtool curl make g++ unzip
+
+## protobuf v3.6.1
+
+git clone https://github.com/protocolbuffers/protobuf
```
Comment (PiotrSikora) [inline WASM.md]: Don't we need your fork?

File: WASM.md [hunk 3]
```diff
@@ -1,23 +1,64 @@
-Proof of Concept WebAssembly Support.
+WebAssembly Extension Support
 
-# WebAssembly VM
+# Dependencies
 
-The WebAssembly VM is https://github.com/WAVM/WAVM. This VM was chosen for the
-proof of concept because:
+## apt-get install libglib2.0-dev (on Debian/Ubuntu).
```
Comment (PiotrSikora) [inline WASM.md]: Nit: split name (glibc headers) from the instructions, like for all other dependencies.

File: WASM.md [hunk 4]
```diff
@@ -1,23 +1,64 @@
...
+However 1.38.25 is known to work.
+
+## clang-7 or clang-8
+
+export CC=clang
+export CXX=clang++
+
+## tools
+
+sudo apt-get install autoconf automake libtool curl make g++ unzip
```
Comment (PiotrSikora) [inline WASM.md]: Those are all normal Envoy dependencies, I don't think we need to replicate "how to build Envoy" instructions here.

File: WASM.md [hunk 5]
```diff
@@ -1,23 +1,64 @@
...
+cd protobuf
+git checkout v3.6.1
+git submodule update --init --recursive
+./autogen.sh
+./configure
+make
+make check
+sudo make install
+
+## WAVM binaries
```
Comment (PiotrSikora) [inline WASM.md]: add "if you want to rebuild WebAssembly tests"

File: WASM.md [hunk 6]
```diff
@@ -1,23 +1,64 @@
-Proof of Concept WebAssembly Support.
+WebAssembly Extension Support
 
-# WebAssembly VM
+# Dependencies
```
Comment (PiotrSikora) [inline WASM.md]: Perhaps split this into "build dependencies" and "dependencies for regenerating tests modules"?

File: WASM.md (hunk 2 above)
Comment (jplevyak) [reply WASM.md]: That is only required if you want to build the .bc files.

File: WASM.md (hunk 1 above)
Comment (jplevyak) [reply WASM.md]: I' will add a note to include clang in your path.

File: WASM.md (hunk 3 above)
Comment (jplevyak) [reply WASM.md]: done.

File: WASM.md (hunk 4 above)
Comment (jplevyak) [reply WASM.md]: done

File: WASM.md (hunk 5 above)
Comment (jplevyak) [reply WASM.md]: done

File: WASM.md (hunk 6 above)
Comment (jplevyak) [reply WASM.md]: done.

**Pull request closed without merging**